File: internal/excel/read_excel.go
package excel

import (
	"onestsignt/internal/codes"
	"onestsignt/internal/normalize"
	"strings"

	"github.com/xuri/excelize/v2"
)

type ReaderExcel struct {
	path string
}

func NewReaderExcel(path string) *ReaderExcel {
	return &ReaderExcel{
		path: path,
	}
}

func (r ReaderExcel) Path() string {
	return r.path
}

func (r ReaderExcel) ReadExcel() (codes.Index, error) {
	file, err := excelize.OpenFile(r.path)
	if err != nil {
		return codes.Index{}, err
	}
	defer file.Close()

	sheets := file.GetSheetList()
	sheetName := sheets[0]

	rows, err := file.GetRows(sheetName)
	if err != nil {
		return codes.Index{}, err
	}

	index := codes.NewIndex()
	diagnostics := codes.Diagnostics{
		File:  r.path,
		Sheet: sheetName,
		Rows:  len(rows),
	}

	if columnIndex, ok := findHeaderColumn(rows, "DM CODE"); ok {
		if err := r.readColumn(rows, sheetName, columnIndex, &index, &diagnostics); err != nil {
			return codes.Index{}, err
		}
		index.SetDiagnostics(diagnostics)
		return index, nil
	}

	for rowIndex, row := range rows {
		for colIndex, cell := range row {
			code := normalize.NormalizeCode(cell)
			if code == "" {
				continue
			}
			diagnostics.NonEmptyCells++

			cellName, err := excelize.CoordinatesToCellName(colIndex+1, rowIndex+1)
			if err != nil {
				return codes.Index{}, err
			}

			location := codes.Location{
				File:  r.path,
				Sheet: sheetName,
				Cell:  cellName,
			}

			if isCodeLike(code) {
				diagnostics.CountedCells++
				index.Add(code, codes.Location{
					File:  r.path,
					Sheet: sheetName,
					Cell:  cellName,
				})
			} else {
				diagnostics.IgnoredNonEmptyCells++
				addIgnoredSample(&diagnostics, location, code, "ячейка не похожа на код маркировки")
			}
		}
	}

	index.SetDiagnostics(diagnostics)
	return index, nil
}

func (r ReaderExcel) readColumn(rows [][]string, sheetName string, columnIndex int, index *codes.Index, diagnostics *codes.Diagnostics) error {
	columnName, err := excelize.ColumnNumberToName(columnIndex + 1)
	if err != nil {
		return err
	}
	diagnostics.Column = columnName

	for rowIndex := 1; rowIndex < len(rows); rowIndex++ {
		row := rows[rowIndex]
		if columnIndex >= len(row) {
			continue
		}

		code := normalize.NormalizeCode(row[columnIndex])
		if code == "" {
			continue
		}

		diagnostics.NonEmptyCells++
		diagnostics.CountedCells++

		cellName, err := excelize.CoordinatesToCellName(columnIndex+1, rowIndex+1)
		if err != nil {
			return err
		}
		index.Add(code, codes.Location{
			File:  r.path,
			Sheet: sheetName,
			Cell:  cellName,
		})
	}

	return nil
}

func findHeaderColumn(rows [][]string, header string) (int, bool) {
	if len(rows) == 0 {
		return 0, false
	}
	header = normalizeHeader(header)
	for index, cell := range rows[0] {
		if normalizeHeader(cell) == header {
			return index, true
		}
	}
	return 0, false
}

func normalizeHeader(value string) string {
	value = strings.TrimSpace(value)
	value = strings.ToUpper(value)
	value = strings.Join(strings.Fields(value), " ")
	return value
}

func isCodeLike(code string) bool {
	if strings.HasPrefix(code, "01") {
		return true
	}
	return len([]rune(code)) >= 20 && strings.Contains(code, "93")
}

func addIgnoredSample(diagnostics *codes.Diagnostics, location codes.Location, value string, reason string) {
	if len(diagnostics.IgnoredSamples) >= 30 {
		return
	}
	diagnostics.IgnoredSamples = append(diagnostics.IgnoredSamples, codes.IgnoredSample{
		Location: location.String(),
		Value:    truncate(value, 80),
		Reason:   reason,
	})
}

func truncate(value string, limit int) string {
	runes := []rune(value)
	if len(runes) <= limit {
		return value
	}
	return string(runes[:limit]) + "..."
}

File: node-api/src/input/read-excel.ts
import XLSX from "xlsx";
import { CodeIndex, stringifyLocation } from "../codes.js";
import { normalizeCode } from "../normalize.js";
import type { Diagnostics, Location } from "../types.js";

export function readExcel(filePath: string): CodeIndex {
  const workbook = XLSX.readFile(filePath, {
    cellDates: false
  });
  const sheetName = workbook.SheetNames[0];
  if (!sheetName) {
    throw new Error("в файле нет листов");
  }
  const sheet = workbook.Sheets[sheetName];
  if (!sheet) {
    throw new Error(`лист "${sheetName}" не найден`);
  }

  const rows = sheetToRows(sheet);
  const index = new CodeIndex();
  const diagnostics: Diagnostics = {
    file: filePath,
    sheet: sheetName,
    rows: rows.length,
    nonEmptyCells: 0,
    countedCells: 0,
    ignoredNonEmptyCells: 0,
    ignoredSamples: []
  };

  const headerColumn = findHeaderColumn(rows, "DM CODE");
  if (headerColumn !== -1) {
    readColumn(filePath, sheetName, rows, headerColumn, index, diagnostics);
    index.setDiagnostics(diagnostics);
    return index;
  }

  rows.forEach((row, rowIndex) => {
    row.forEach((cell, columnIndex) => {
      const code = normalizeCode(cell);
      if (code === "") {
        return;
      }
      diagnostics.nonEmptyCells += 1;

      const location: Location = {
        file: filePath,
        sheet: sheetName,
        cell: XLSX.utils.encode_cell({ r: rowIndex, c: columnIndex })
      };

      if (isCodeLike(code)) {
        diagnostics.countedCells += 1;
        index.add(code, location);
      } else {
        diagnostics.ignoredNonEmptyCells += 1;
        addIgnoredSample(diagnostics, location, code, "ячейка не похожа на код маркировки");
      }
    });
  });

  index.setDiagnostics(diagnostics);
  return index;
}

function sheetToRows(sheet: XLSX.WorkSheet): string[][] {
  const rawRows = XLSX.utils.sheet_to_json(sheet, {
    header: 1,
    defval: "",
    raw: false,
    blankrows: false
  }) as unknown[][];

  return rawRows.map((row) => {
    if (!Array.isArray(row)) {
      return [];
    }
    return row.map((cell) => (cell == null ? "" : String(cell)));
  });
}

function readColumn(
  filePath: string,
  sheetName: string,
  rows: string[][],
  columnIndex: number,
  index: CodeIndex,
  diagnostics: Diagnostics
): void {
  diagnostics.column = XLSX.utils.encode_col(columnIndex);

  for (let rowIndex = 1; rowIndex < rows.length; rowIndex += 1) {
    const row = rows[rowIndex];
    if (!row || columnIndex >= row.length) {
      continue;
    }

    const code = normalizeCode(row[columnIndex] ?? "");
    if (code === "") {
      continue;
    }

    diagnostics.nonEmptyCells += 1;
    diagnostics.countedCells += 1;
    index.add(code, {
      file: filePath,
      sheet: sheetName,
      cell: XLSX.utils.encode_cell({ r: rowIndex, c: columnIndex })
    });
  }
}

function findHeaderColumn(rows: string[][], header: string): number {
  const firstRow = rows[0];
  if (!firstRow) {
    return -1;
  }

  const normalizedHeader = normalizeHeader(header);
  return firstRow.findIndex((cell) => normalizeHeader(cell) === normalizedHeader);
}

function normalizeHeader(value: string): string {
  return value.trim().toUpperCase().split(/\s+/).filter(Boolean).join(" ");
}

function isCodeLike(code: string): boolean {
  if (code.startsWith("01")) {
    return true;
  }
  return [...code].length >= 20 && code.includes("93");
}

function addIgnoredSample(
  diagnostics: Diagnostics,
  location: Location,
  value: string,
  reason: string
): void {
  if (diagnostics.ignoredSamples.length >= 30) {
    return;
  }
  diagnostics.ignoredSamples.push({
    location: stringifyLocation(location),
    value: truncate(value, 80),
    reason
  });
}

function truncate(value: string, limit: number): string {
  const runes = [...value];
  if (runes.length <= limit) {
    return value;
  }
  return `${runes.slice(0, limit).join("")}...`;
}
